core/src/main/java/com/linecorp/armeria/internal/server/annotation/AnnotatedValueResolver.java

```diff
@@ -621,6 +634,46 @@ private static AnnotatedValueResolver ofRequestObject(String name, AnnotatedElem
                 .build();
     }
 
+    private static AnnotatedValueResolver ofAttribute(String name,
+                                                      Attribute attr,
+                                                      AnnotatedElement annotatedElement,
+                                                      AnnotatedElement typeElement, Class<?> type,
+                                                      DescriptionInfo description) {
+
+        final List<AttributeKey<?>> attrKeys = new ArrayList<>();
+
+        final AttributeKey<?> prefixTypeAttrKey = AttributeKey.valueOf(attr.prefix(), name);
+        attrKeys.add(prefixTypeAttrKey);
+
+        final AttributeKey<?> declaredTypeKey = AttributeKey.valueOf(type, name);
+        attrKeys.add(declaredTypeKey);
+
+        if (type.isPrimitive()) {
+            final Class<?> mustRefType = Primitives.wrap(type);
+            final AttributeKey<?> mustRefTypeKey = AttributeKey.valueOf(mustRefType, name);
+            attrKeys.add(mustRefTypeKey);
+        }
+
+        final AttributeKey<Object> objectTypeAttrKey = AttributeKey.valueOf(name);
+        attrKeys.add(objectTypeAttrKey);
+
+        return new Builder(annotatedElement, type, name)
+                .annotationType(Attribute.class)
+                .typeElement(typeElement)
+                .supportDefault(true)
+                .supportContainer(true)
+                .description(description)
+                .resolver(attributeResolver(ctx -> {
+                                                for (AttributeKey<?> attrKey : attrKeys) {
+                                                    final Object value = ctx.context.attr(attrKey);
+                                                    if (value != null)
+                                                        return value;
+                                                }
+                                                return null;
+                }))
```

Comment:
We could simplify this by just passing the list of `AttributeKey`s:

```
.resolver(attributeResolver(attrKeys))
```

Then `attributeResolver()` could find the attribute by itself.

Reply:
i see! i applied it.